### PR TITLE
Delay evaluation of ExtraArguments and standardize puppet-server and puppet-masterless

### LINKED ISSUE #4462: provisioner/puppet: make execute_command variables available to extra_arguments
extra_arguments is a perfect place to reference class variables but it's interpolated so early NULLs get substituted instead. Here is example:
`{{if ne ".ModulePath" ""}}--modulepath='{{.ModulePath}}:$basemodulepath'{{end}}`

The specific example is a result of Puppet's implementation choice whereby '--modulepath' obliterates all of the standard module paths and until such time as they implement an "additive" mode, the values have to be restored. Sure one could completely rewrite "execute_command" but why not just fix the problem?

Conceivably one could concoct an "execute_command" where one of the extra's is a subshell invocation of a command that looks something like
`"--modulepath=$( {{.PuppetBinDir}}/puppet config print modulepath --environment={{user `environment`}} )`

Early evaluation means PuppetBinDir would likewise been rendered as NULL.

### PR DESCRIPTION
**Pull request by tb3088**
Replaces #4488 which is dirty.

Fixes pre-mature interpolation of "extra_arguments" and adds "extra_arguments" to puppet-server provisioner. This also brings puppet-server/masterless much closer to each other for argument and program flow consistency.

Here is an example from my template:
```
"extra_arguments" : [
"{{if ne "{{user environment}}" ""}}--environment={{user environment}}{{end}}",
"{{if user noop}}--noop{{end}}",
"{{if ne ".ModulePath" ""}}--modulepath='{{.ModulePath}}:$basemodulepath'{{end}}"
```

Closes #4462

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,2 +1,3 @@
 
-common/test-fixtures/root/*	eol=lf
\ No newline at end of file
+common/test-fixtures/root/*	eol=lf
+*.go	eol=lf
```
Comment (SwampDragons) [inline .gitattributes]: Is this relevant to the rest of the PR or left in from something else?

File: provisioner/puppet-masterless/provisioner.go [hunk 2]
```diff
@@ -26,6 +26,9 @@ type Config struct {
 	// Additional arguments to pass when executing Puppet
 	ExtraArguments []string `mapstructure:"extra_arguments"`
 
+	// The Guest OS Type (unix or windows)
+	GuestOSType string `mapstructure:"guest_os_type"`
```
Comment (SwampDragons) [inline provisioner/puppet-masterless/provisioner.go]: Can we make this ordering alphabetical?

File: provisioner/puppet-masterless/provisioner.go [hunk 3]
```diff
@@ -62,45 +65,48 @@ type Config struct {
...
 type guestOSTypeConfig struct {
+	tempDir          string
 	stagingDir       string
 	executeCommand   string
 	facterVarsFmt    string
 	facterVarsJoiner string
 	modulePathJoiner string
 }
 
+// FIXME assumes both Packer host and target are same OS
```
Comment (SwampDragons) [inline provisioner/puppet-masterless/provisioner.go]: Will this code break if this assumption is not met?

Comment (SwampDragons) [reply provisioner/puppet-masterless/provisioner.go]: You can find out the host system by using runtime.GOOS (you can search the code for examples)

File: provisioner/puppet-masterless/provisioner.go [hunk 4]
```diff
@@ -162,10 +170,6 @@ func (p *Provisioner) Prepare(raws ...interface{}) error {
 		p.config.ExecuteCommand = p.guestOSTypeConfig.executeCommand
 	}
 
-	if p.config.ExecuteCommand == "" {
```
Comment (SwampDragons) [inline provisioner/puppet-masterless/provisioner.go]: Good catch.

File: provisioner/puppet-masterless/provisioner.go [hunk 5]
```diff
@@ -296,8 +299,16 @@ func (p *Provisioner) Provision(ui packer.Ui, comm packer.Communicator) error {
 		PuppetBinDir:    p.config.PuppetBinDir,
 		Sudo:            !p.config.PreventSudo,
 		WorkingDir:      p.config.WorkingDir,
-		ExtraArguments:  strings.Join(p.config.ExtraArguments, " "),
+		ExtraArguments:  "",
+	}
+
+	p.config.ctx.Data = &data
+	_ExtraArguments, err := interpolate.Render(strings.Join(p.config.ExtraArguments, " "), &p.config.ctx)
```
Comment (SwampDragons) [inline provisioner/puppet-masterless/provisioner.go]: So what benefit do you gain from interpolating the ExtraArguments separately instead of at the same time? Casually reading this section of code it looks like the outcome will be the same.

File: provisioner/puppet-server/provisioner.go [hunk 6]
```diff
@@ -77,16 +43,17 @@ type Config struct {
...
 	// If true, `sudo` will NOT be used to execute Puppet.
 	PreventSudo bool `mapstructure:"prevent_sudo"`
 
 	// The directory where files will be uploaded. Packer requires write
 	// permissions in this directory.
 	StagingDir string `mapstructure:"staging_dir"`
 
+	// The directory from which the command will be executed.
+	// Packer requires the directory to exist when running puppet.
+	WorkingDir string `mapstructure:"working_directory"`
```
Comment (SwampDragons) [inline provisioner/puppet-server/provisioner.go]: Again, I know it's a bikeshed but I'd like these fields to be alphabetical.

File: provisioner/puppet-masterless/provisioner.go [hunk 7]
```diff
@@ -62,45 +65,48 @@ type Config struct {
...
+// FIXME assumes both Packer host and target are same OS
 var guestOSTypeConfigs = map[string]guestOSTypeConfig{
 	provisioner.UnixOSType: {
+		tempDir:    "/tmp",
 		stagingDir: "/tmp/packer-puppet-masterless",
 		executeCommand: "cd {{.WorkingDir}} && " +
 			`{{if ne .FacterVars ""}}{{.FacterVars}} {{end}}` +
 			"{{if .Sudo}}sudo -E {{end}}" +
 			`{{if ne .PuppetBinDir ""}}{{.PuppetBinDir}}/{{end}}` +
-			`puppet apply --verbose --modulepath='{{.ModulePath}}' ` +
```
Comment (SwampDragons) [inline provisioner/puppet-masterless/provisioner.go]: What's your reasoning behind removing the verbose and modulepath flags?

File: .gitattributes (hunk 1 above)
Comment (tb3088) [reply .gitattributes]: I have a separate PR #4494 that deals with a lot more EOL/Windows issues. Technically no EOL is an error even if the *.go line is gratuitous. reverted.

File: provisioner/puppet-masterless/provisioner.go (hunk 2 above)
Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: Happy to have it alphabetical. But the primary motivation was to normalize the section across puppet-masterless and puppet-server.

File: provisioner/puppet-masterless/provisioner.go (hunk 3 above)
Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: Yes it will break. Even Windows->Windows will break since SYSTEMRoot on local and remote can easily be different. runtime.GOOS only tells me the OS of the machine running Packer, not what's on the other end of the provisioner. Am I wrong?

Comment (SwampDragons) [reply provisioner/puppet-masterless/provisioner.go]: No, but GuestOSType will tell us what is on the other side of the provisioner. So using them together should give you the full picture, right?

Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: it helps but 2 windows instances can have different %SYSTEMROOT% values. 'stagingDir' is particularly problematic. I'm pretty new at this. Not sure how to 'query' system attributes of the remote end. Even in Unix-land (/tmp) is not necessarily assured and we don't actually test existence nor writability/execute permissions.

File: provisioner/puppet-masterless/provisioner.go (hunk 7 above)
Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: I didn't remove modulepath, just wrapped it in an IF. So that if the template omitted it, the defaults would get used. Furthermore, I do things like:

``{{if ne ".ModulePath" ""}}--modulepath="{{.ModulePath}}:$({{.PuppetBinDir}}/puppet config print {{if ne "{{user `environment`}} ""}}--environment={{user `environment`}}{{end}} modulepath)'{{end}}"`

IMO '--verbose' is entirely too wordy and useful information gets lost in the deluge of output.

File: provisioner/puppet-masterless/provisioner.go (hunk 2 above)
Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: FWIW pretty much none of the other provisioners sort their Config structs. :)

Comment (SwampDragons) [reply provisioner/puppet-masterless/provisioner.go]: yeah, but adding a little code hygiene when you make changes never hurt anyone. I'm not going to block your PR on this, it's just nice to do

File: provisioner/puppet-masterless/provisioner.go (hunk 3 above)
Comment (SwampDragons) [reply provisioner/puppet-masterless/provisioner.go]: Yeah, it's an interesting problem. With most of the provisioners we try not to know anything about the guest system, use reasonable defaults/env vars where we can, and make sure that potential problem areas are configurable.

It may be worth it in the future to think about allowing users to directly customize the entire guestOSTypeConfig, though that's outside scope for this PR.

It looks to me like this issue already existed before you PR, and you're just adding a comment explaining that it's a problem. Is that accurate?

File: provisioner/puppet-masterless/provisioner.go (hunk 7 above)
Comment (SwampDragons) [reply provisioner/puppet-masterless/provisioner.go]: Ah, now I see it. That seems like a good change for modulepath.

As for the --verbose... I have no strong opinion either way; just wanted to know your reasoning for the change. In general I tend to agree that it's probably spammy. Go ahead and keep it out.

**Comment by SwampDragons** [comment]
Okay, I think this is ready for testing. Do you know any community members who'd be willing to give this patch a try? I'd like to get some real people using this on their use cases.

File: provisioner/puppet-masterless/provisioner.go (hunk 3 above)
Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: yes it was just adding a heads-up comment.

File: provisioner/puppet-masterless/provisioner.go (hunk 2 above)
Comment (tb3088) [reply provisioner/puppet-masterless/provisioner.go]: datastructures alphabetized

**Comment by tb3088** [comment]
unfortunately I don't have such a Rolodex. Delayed evaluation of Extra_args is a use case that apparently doesn't occur to much of the user base which only does trivial things. (I blame the documentation. *grin*)

**Comment by SwampDragons** [comment]
lol well in that case I'll try to get an environment up and running to test this soon.  And if you have docs improvements, I promise those PRs will get merged faster than this one!

**Comment by SwampDragons** [comment]
Might also be good to add a note to the `extra_arguments` documentation saying that the template variables used in execute_command are also available to `extra_arguments`

**Comment by tb3088** [comment]
documentation updated @SwampDragons
